File: src/main/kotlin/shortestpath/개선된_다익스트라_알고리즘.java
package shortestpath;

import java.util.*;
import java.math.*;

public class 개선된_다익스트라_알고리즘 {

    private static int[] distances;
    private static List<List<Node>> graph = new ArrayList<>();

    private static final int INF = (int)1e9;

    public static void dijkstra(int start){
        PriorityQueue<Node> pq = new PriorityQueue<>();
        //시작노드 삽입 및 최단거리 0으로 초기화
        pq.offer(new Node(start, 0));
        distances[start] = 0;

        //큐가 빌 때까지 반복
        while(!pq.isEmpty()){
            Node now = pq.poll();
            int index = now.getIndex();
            int distance = now.getDistance();
            //이미 처리된 노드라면 무시
            if(distances[index] < distance) continue;
            //현재 노드와 연결된 노드 처리
            for(int i = 0; i < graph.get(index).size(); i++){
                int cost = distance + graph.get(index).get(i).getDistance();
                //더 저렴하다면 업데이트
                if(cost < distances[graph.get(index).get(i).getIndex()]){
                    pq.offer(new Node(graph.get(index).get(i).getIndex(), cost));
                    distances[graph.get(index).get(i).getIndex()] = cost;
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        //노드의 개수와 간선의 개수 입력받기
        int n = sc.nextInt();
        int m = sc.nextInt();
        //시작노드 입력받기
        int start = sc.nextInt();
        //최단거리 테이블 모두 무한대로 초기화
        distances = new int[n+1];
        for(int i = 0; i <= n; i++){
            distances[i] = INF;
        }

        //그래프 초기화 및 입력받기
        for(int i = 0; i <= n; i++){
            graph.add(new ArrayList<>());
        }
        for(int i = 0; i < m; i++){
            int a = sc.nextInt();
            int b = sc.nextInt();
            int distance = sc.nextInt();

            graph.get(a).add(new Node(b, distance));
        }
        //다익스트라 최단경로 알고리즘 수행
        dijkstra(start);

        for(int i = 1; i <= n; i++){
            if(distances[i] == INF) System.out.println("INFINITY");
            else System.out.println(distances[i]);
        }
    }

    static class Node implements Comparable<Node> {
        private int index;
        private int distance;

        public Node(int index, int distance){
            this.index = index;
            this.distance = distance;
        }

        public int getIndex(){
            return this.index;
        }

        public int getDistance(){
            return this.distance;
        }

        //거리가 짧은 순서대로 정렬
        @Override
        public int compareTo(Node other){
            return this.distance - other.distance;
        }
    }
}

File: src/main/kotlin/shortestpath/개선된_다익스트라_알고리즘.kt
package shortestpath

import java.util.*
import kotlin.Comparator
import kotlin.math.*

private const val INF = 1e9.toInt()

fun main(){
    val sc = Scanner(System.`in`)
    //노드의 개수와 간선의 개수 입력받기
    val n = sc.nextInt()
    val m = sc.nextInt()
    //시작노드 번호 입력받기
    val start = sc.nextInt()
    //최단거리 테이블 초기화
    val distances = IntArray(n + 1) { INF }
    //그래프 입력받기
    val graph = mutableListOf<MutableList<Node>>()
    for(i in 0 .. n){
        graph.add(mutableListOf())
    }
    for(i in 0 until m){
        val a = sc.nextInt()
        val b = sc.nextInt()
        val cost = sc.nextInt()

        graph[a].add(Node(b, cost))
    }

    fun dijkstra(start: Int){
        val pq = PriorityQueue<Node>()
        //시작 노드는 최단 경로를 0으로 설정
        pq.add(Node(start, 0))
        distances[start] = 0
        //큐가 빌 때까지 반복
        while(!pq.isEmpty()){
            val node = pq.poll()
            val now = node.index
            val distance = node.distance
            //현재 노드가 이미 처리된 적이 있다면 무시
            if(distances[now] < distance) continue
            //현재 노드와 연결된 다른 인접한 노드들을 확인
            for(i in 0 until graph[now].size){
                val cost = distance + graph[now][i].distance
                //비용이 더 적게 든다면
                if(cost < distances[graph[now][i].index]){
                    distances[graph[now][i].index] = cost
                    pq.offer(Node(graph[now][i].index, cost))
                }
            }
        }
    }

    dijkstra(start)

    for(i in 1 .. n){
        when(distances[i]){
            INF -> println("INFINITY")
            else -> println(distances[i])
        }
    }
}

data class Node(val index: Int, val distance: Int): Comparable<Node> {
    override fun compareTo(other: Node): Int = this.distance - other.distance
}